ddr/lib/ddr-scanner/dwarf/DwarfScanner.cpp

```diff
@@ -1647,17 +1647,26 @@ DwarfScanner::getSuperUDT(Dwarf_Die die, ClassUDT *udt)
 	DDR_RC rc = DDR_RC_ERROR;
 	Dwarf_Half tag = 0;
 	Dwarf_Die superTypeDie = NULL;
-
-	if (DDR_RC_OK == getTypeTag(die, &superTypeDie, &tag)) {
-		ClassUDT *superUDT = NULL;
-		/* Get the super udt. */
-		if ((DDR_RC_OK == addDieToIR(superTypeDie, tag, NULL, (Type **)&superUDT))
-				&& (NULL != superUDT)) {
-			rc = DDR_RC_OK;
-			udt->_superClass = superUDT;
+	Dwarf_Die currentDie = die;
+	ClassUDT *superUDT = NULL;
+	bool iterating = false;
+	do {
+		tag = 0;
+		if (DDR_RC_OK == getTypeTag(currentDie, &superTypeDie, &tag)) {
+			/* Get the super udt. */
+			rc = addDieToIR(superTypeDie, tag, NULL, (Type **)&superUDT);
+			currentDie = superTypeDie;
+			iterating = true;
+		} else if (not iterating) {
+			/* No super type found. */
+			return rc;
```

Comment:
This highlights the unreasonable cast to `(Type **)` - it was wrong before and it's still wrong. As I understand it, this change is motivated exactly because `superUDT` is not a class, but a typedef.

The proper way to proceed is to:
* change `ClassUDT`:
```c
	Type *_superClass;
````
* change the local here:
```c
		Type *superUDT = NULL;
```
* to remove that bad cast:
```c
		if ((DDR_RC_OK == addDieToIR(superTypeDie, tag, NULL, &superUDT))
```
* to update `MergeVisitor::visitClass`
```c
		Type *otherSuper = ((ClassUDT *)_other)->_superClass;
...
				otherSuper = thisSuper;
```
There are general mechanisms to hide typedef types, so this may be all that is necessary.

Reply:
Looks like the above suggestions compile fine (atleast for Open XL), will push and check with Jenkins

Reply:
Please revert the loop here; this is not the right place to deal with typedefs.